Stepper: Rename the SiteCreationStep to CreateSite

markbiek opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes #71707

## Proposed Changes

For clarity (as discussed [here](https://github.com/Automattic/wp-calypso/pull/71611#discussion_r1062397447)), this renames the `SiteCreationStep` to `CreateSite`

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Apply this branch locally or use the calypso.live url.
* Make sure you're seeing Tracks events in the dev console. When the `create-site` step starts, you should see tracks events for both `create-site` _and_ `site-creation-step`.
![image](https://github.com/Automattic/wp-calypso/assets/917632/b9804b9d-19a3-4f73-bcba-bdc1fbceac89)

* The following flows need to be tested. I checked off the flows that I knew how to test. The others I wasn't sure about.

- [x] ai-assembler
- [x] assembler-first-flow
- [ ] connect-domain
- [ ] copy-site
- [ ] design-first
- [ ] free
- [ ] hundred-year-plan
- [ ] import-flow
- [x] import-hosted-site
- [ ] link-in-bio
- [ ] new-hosted-site-flow
- [ ] newsletter
- [ ] reblogging
- [ ] start-writing
- [ ] tailored-ecommerce-flow
- [ ] trial-woo-express-flow

Here are the flows that I tested myself:
- [x] ai-assembler
- [x] assembler-first-flow
- [x] free
- [x] link-in-bio
- [x] newsletter
- [x] tailored-ecommerce-flow
- [x] trial-woo-express-flow

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

markbiek committed:
Move site-creation-step to create-site

markbiek committed:
Update flows to use create-site

sixhours commented:
I presume the best way to test this is to run through each flow with this branch and ensure that they all create a site. I'll check each one off as I go:

- [ ] ai-assembler
- [ ] assembler-first-flow
- [ ] connect-domain
- [ ] copy-site
- [ ] design-first
- [ ] free
- [ ] hundred-year-plan
- [ ] import-flow
- [ ] import-hosted-site
- [ ] link-in-bio
- [ ] new-hosted-site-flow
- [ ] newsletter
- [ ] reblogging
- [ ] start-writing
- [ ] tailored-ecommerce-flow
- [ ] trial-woo-express-flow

markbiek commented:
> I presume the best way to test this is to run through each flow with this branch and ensure that they all create a site. 

@sixhours That's how I did it. Thanks for testing!

sixhours commented:
I've been able to test every flow except the two import flows; I couldn't figure out where to go to trigger those. Looks like maybe @Automattic/caribou might know! 🤞

markbiek commented:
> I've been able to test every flow except the two import flows; I couldn't figure out where to go to trigger those. 

Just for fun, I tried going to `/setup/import-focused` and that did take me to a stepper flow. I was prompted to enter the url of a site to import but then it seemed to hang on a loading screen forever so there must be more to the flow.

However I was able to get `/setup/import-hosted-site` to work! I didn't get all the way through the import process but it definitely created a new site as part of the flow.

ouikhuan reviewed (commented):
> I've been able to test every flow except the two import flows; I couldn't figure out where to go to trigger those. Looks like maybe @Automattic/caribou might know! 🤞

Hi folks, thanks for the ping! I've tested both `import-flow` and `import-hosted-site` flow and they both works as expected.

- [x] import-flow
- [x] import-hosted-site

If you're curious about how to trigger the create site action on these two flows, I've recorded step to step videos below:

**Import flow:**

- Start with a self-hosted WordPress site and install `Move to WordPress.com` plugin.
- To save you some time you can start directly from `http://calypso.localhost:3000/setup/import-focused/migrationHandler?from=SELF_HOSTED_SITE_SLUG`

https://github.com/Automattic/wp-calypso/assets/4074459/acf607cc-ae00-4bd9-8c7d-b460991e65da

**Import hosted site:**

- Start with https://wordpress.com/hosting/ and click `Start for free` CTA
- To save you some time you can start directly from `http://calypso.localhost:3000/sites?hosting-flow=true` and click `Migrate a site` CTA


https://github.com/Automattic/wp-calypso/assets/4074459/c7c56e12-a036-4670-b4ce-ca4c5c2317c4

Hope this helps!

daledupreez reviewed (commented):
I have a key concern with making this change: does the change impact the Tracks events we create, especially `calypso_signup_step_start`? I am worried this could break or otherwise interfere with key analysis for signup flows.

markbiek commented:
> I have a key concern with making this change: does the change impact the Tracks events we create, especially `calypso_signup_step_start`? I am worried this could break or otherwise interfere with key analysis for signup flows.

@daledupreez I went through `/setup/free` with Tracks Vigilante on both this branch and production and, as far as I can tell, the signup-specific tracks events are the same.

The only place I'm seeing a change is on some `calypso_page_view` events, the path property changes from ` /setup/free/site-creation-step` to `/setup/free/create-site`.

daledupreez commented:
@markbiek, I believe the top-level events are the same, but the `step` event prop looks like it has changed to reflect the new step slug. This means any funnels built with a step that looks for `calypso_signup_step_start` events with `step` = `site-creation-step` will stop working as expected.

markbiek commented:
I'm a little stuck on how to proceed with this without messing up the Tracks events.

Here are the options that I see:

1. Change the string `create-site` back to `site-creation-step` (so the Tracks events don't change) but leave the change to the component/step name. But then, is it worth changing the component name?
2. Figure out some other way to have the Tracks events get recorded with the old name. But I don't think there's a good way to do this.
3. Abandon the change.

daledupreez commented:
@markbiek, could we log Tracks events with both `step` values? That would allow future uses to rely on the new name, and older users to continue using the existing name. It would be "noisy", but would involve minimal downstream changes that could end up being risky.

markbiek commented:
> @markbiek, could we log Tracks events with both `step` values? That would allow future uses to rely on the new name, and older users to continue using the existing name. It would be "noisy", but would involve minimal downstream changes that could end up being risky.

The only issue I see with that would be that we'd have to add a special exception (probably [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/internals/index.tsx#L166)). We'd have to check for `create-site` as the current step name and then also manually record an event for `site-creation-step`.

markbiek committed:
Record tracks events for a step's old slug

When we call `recordStepStart` to record a tracks event for a step, we
check to see if the step has an entry in `getStepOldSlug`. If it does,
we fire an additional tracks event for the old slug.

markbiek commented:
OK, I've updated with an approach that logs tracks events with both `site-creation-step` and `create-site`.

![image](https://github.com/Automattic/wp-calypso/assets/917632/78c98b5b-6fcc-4d6d-a3d3-2d685bb0cada)

I added an optional `old_slug` property to the Step definition. All flows that use the `create-site` step have an `old_slug` of `site-creation-step`. For those flows, we call `recordStepStart` twice (for the slug and old_slug).

I'm not totally convinced this is the best way to go about it so any feedback is welcome. The biggest issue is that someone has to know to add the `old_slug` to the create-site step in their flow. I worry that it will be easy to miss if new flows are created.

Another, possibly better, way to go might be to have a global map of new site slugs to old site-slugs. Then the data is only in one place and people making new flows don't have to worry about anything.

markbiek commented:
I've been thinking about it and I've convinced myself that adding a more global map of tracks events is the way to go, and not the way I'm currently doing it.

So ignore the most recent changes as I'll be replacing them soon.

markbiek commented:
Updated using a map approach instead of my original idea.

daledupreez commented on client/landing/stepper/declarative-flow/internals/index.tsx:
```diff
@@ -62,8 +62,20 @@ export const FlowRenderer: React.FC< { flow: Flow } > = ( { flow } ) => {
+			'create-site': 'site-creation-step',
+		};
+
+		return stepSlugMap[ stepSlug ];
```
Neatly done!

daledupreez commented on client/landing/stepper/declarative-flow/link-in-bio-domain.ts:
```diff
@@ -54,8 +54,8 @@ const linkInBioDomain: Flow = {
 			{
-				slug: 'siteCreationStep',
-				asyncComponent: () => import( './internals/steps-repository/site-creation-step' ),
+				slug: 'createSiteStep',
```
This should be `createSite` so we get the correct mapping.
```suggestion
				slug: 'createSite',
```

daledupreez commented on client/landing/stepper/declarative-flow/link-in-bio-domain.ts:
```diff
@@ -111,10 +111,10 @@ const linkInBioDomain: Flow = {
 					}
-					return navigate( 'siteCreationStep' );
+					return navigate( 'createSiteStep' );
 				}
 
-				case 'siteCreationStep':
+				case 'createSiteStep':
```
As above, these should be `createSite`.

daledupreez commented on client/landing/stepper/declarative-flow/link-in-bio.ts:
```diff
@@ -98,9 +98,9 @@ const linkInBio: Flow = {
 
-				case 'siteCreationStep':
+				case 'createSiteStep':
 					return navigate( 'processing' );
```
These should all be `createSite`, not `createSiteStep`.

daledupreez commented on client/landing/stepper/declarative-flow/new-hosted-site-flow.ts:
```diff
@@ -66,14 +66,14 @@ const hosting: Flow = {
 				}
 
-				case 'siteCreationStep':
+				case 'createSiteStep':
```
These should all be `createSite`, not `createSiteStep`.

daledupreez reviewed (changes requested):
Hmmm... We either need to standardise on the new step names/slugs, or we need to add more mappings. Currently, we only have mappings for the `create-site` slug, but there are a number of `createSiteStep` slugs in the PR.

markbiek committed:
Change createSiteStep to createSite for consistency

gabrielcaires commented on client/landing/stepper/declarative-flow/ai-assembler.ts:
```diff
@@ -148,20 +148,20 @@ const withAIAssemblerFlow: Flow = {
 					if ( providedDependencies?.filteredSitesCount === 0 ) {
 						// No unlaunched sites, redirect to new site creation step
-						return navigate( 'site-creation-step' );
+						return navigate( 'create-site' );
```
Can we move it to a constant or method to avoid errors caused by typos?

markbiek replied:
I'm not opposed to the idea but it's worth noting that we don't currently do this for any of the steps.

I think it would be better to tackle that as a separate PR where we set up a common area for those constants.

sixhours commented:
I'm seeing two events for `calypso_signup_step_start` with both `create-site` and `site-creation-step` as the `step` property, so that's good. 👍 

<img width="1218" alt="Screenshot 2024-02-05 at 12 25 57 PM" src="https://github.com/Automattic/wp-calypso/assets/2124984/904e26bd-f415-4794-aba8-b78aaab4ca04">

I also noticed there's a `calypso_signup_actions_submit_step` event with only `create-site` for the `step` value. Is that expected?

Do we need to account for the `calypso_signup_actions_submit_step` event with the old `step` value as well or is that not used in our existing funnels/stats?

markbiek commented:
> Do we need to account for the `calypso_signup_actions_submit_step` event with the old `step` value as well or is that not used in our existing funnels/stats?

@sixhours I've done a bunch of digging in Superset and I don't see any existing funnels that use the old `step` value.

Superset query: 336b3-pb/#plain

Results: 336b4-pb/#sql

daledupreez commented on client/landing/stepper/declarative-flow/tailored-ecommerce-flow.ts:
```diff
@@ -22,6 +22,14 @@ import { USER_STORE, ONBOARD_STORE, SITE_STORE } from '../stores';
 import getQuantityFromStorageType from '../utils/get-quantity-from-storage-slug';
 import { getLoginUrl } from '../utils/path';
 import { recordSubmitStep } from './internals/analytics/record-submit-step';
+import CheckPlan from './internals/steps-repository/check-plan';
+import CreateSite from './internals/steps-repository/create-site';
+import DesignCarousel from './internals/steps-repository/design-carousel';
+import DomainsStep from './internals/steps-repository/domains';
+import ProcessingStep from './internals/steps-repository/processing-step';
+import StoreProfiler from './internals/steps-repository/store-profiler';
+import WaitForAtomic from './internals/steps-repository/wait-for-atomic';
+import WaitForPluginInstall from './internals/steps-repository/wait-for-plugin-install';
```
This looks like it may have been from a poor rebase/merge -- I think these imports are no longer used due to us loading the steps asynchronously.
```suggestion
```

daledupreez commented on client/landing/stepper/declarative-flow/trial-wooexpress-flow.ts:
```diff
@@ -10,7 +10,12 @@ import { USER_STORE, ONBOARD_STORE, SITE_STORE } from '../stores';
 import { getLoginUrl } from '../utils/path';
 import { recordSubmitStep } from './internals/analytics/record-submit-step';
 import { AssignTrialResult } from './internals/steps-repository/assign-trial-plan/constants';
+import CreateSite from './internals/steps-repository/create-site';
+import ErrorStep from './internals/steps-repository/error-step';
+import ProcessingStep from './internals/steps-repository/processing-step';
 import { ProcessingResult } from './internals/steps-repository/processing-step/constants';
+import WaitForAtomic from './internals/steps-repository/wait-for-atomic';
+import WaitForPluginInstall from './internals/steps-repository/wait-for-plugin-install';
```
This looks like another rebase/merge issue -- I don't think we need these imports as we're loading the step components asynchronously.
```suggestion
import { ProcessingResult } from './internals/steps-repository/processing-step/constants';
```

daledupreez reviewed (commented):
I've just tested and I am not seeing the second (i.e. the original) event for the `link-in-bio` flow. Not sure what's going wrong here, or why that might not be working.

markbiek committed:
Remove unneeded imports

markbiek commented:
> > Do we need to account for the `calypso_signup_actions_submit_step` event with the old `step` value as well or is that not used in our existing funnels/stats?
> 
> @sixhours I've done a bunch of digging in Superset and I don't see any existing funnels that use the old `step` value.
> 
> Superset query: 336b3-pb/#plain
> 
> Results: 336b4-pb/#sql

Actually, this may not be accurate. I ran another query against `calypso_signup_step_start` and I don't see either `create-site` or `site-creation-step` in the results. Seems like that Superdata data may not be current.

markbiek commented:
That unit test failure is real. I'm looking at that and the `link-in-bio` thing now.

markbiek committed:
Track old step names on the calypso_signup_actions_submit_step event

sixhours reviewed (approved):
LGTM!
